cmd/clusterctl/pkg/client/config/providers_client.go

```diff
@@ -0,0 +1,202 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"net/url"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/klog"
+	clusterctlv1 "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
+)
+
+const (
+	CAPI               = "cluster-api"
+	ProvidersConfigKey = "providers"
+)
+
+// ProvidersClient has methods to work with provider configurations.
+type ProvidersClient interface {
+	// List returns all the provider configurations, including provider configurations hard-coded in clusterctl
+	// and user-defined provider configurations read from the clusterctl configuration file.
+	// In case of conflict, user-defined provider override the hard-coded configurations.
+	List() ([]Provider, error)
+
+	// Get return the configuration for the provider with a given name.
+	// In case the name does not correspond to any existing provider, an error is returned.
+	Get(name string) (*Provider, error)
+}
+
+// providersClient implements ProvidersClient.
+type providersClient struct {
+	reader Reader
+}
+
+// ensure providersClient implements ProvidersClient.
+var _ ProvidersClient = &providersClient{}
+
+func newProvidersClient(reader Reader) *providersClient {
+	return &providersClient{
+		reader: reader,
+	}
+}
+
+func (p *providersClient) defaults() []Provider {
+
+	// clusterctl includes an hard-coded list of "default" providers in order to provide to the users the simplest
+	// out-of-box experience. Other providers can be added by using the clusterctl configuration file.
+
+	// if you are a developer of a community maintained provider, you can send a PR to extend the following list.
+
+	defaults := []Provider{
+		// cluster API core provider
+		&provider{
+			name:  CAPI,
+			url:   "https://github.com/kubernetes-sigs/cluster-api/releases/latest/cluster-api-components.yaml",
+			ttype: clusterctlv1.CoreProviderType,
+		},
+
+		// Infrastructure providersClient
+		&provider{
+			name:  "aws",
+			url:   "https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases/latest/infrastructure-components.yaml",
+			ttype: clusterctlv1.InfrastructureProviderType,
+		},
+		&provider{
+			name:  "docker",
+			url:   "https://github.com/kubernetes-sigs/cluster-api-provider-docker/releases/latest/provider_components.yaml",
+			ttype: clusterctlv1.InfrastructureProviderType,
+		},
+		&provider{
+			name:  "vsphere",
+			url:   "https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/releases/latest/infrastructure-components.yaml",
+			ttype: clusterctlv1.InfrastructureProviderType,
+		},
+
+		// Bootstrap providersClient
```

Comment:
If we are expecting initial usage with v1alpha3, then we should remove references to the CAPBK repo.

Reply:
if you don't mind I will keep CAPBK around for some time more because testing with a v1alpha2 now is far easier than testing with v1alpha3 (still WIP). 

I have added a TODO to remind me of this task.